src/Native/Runtime/thread.cpp

```diff
@@ -274,6 +274,68 @@ PTR_ExInfo Thread::GetCurExInfo()
 /////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 #ifndef DACCESS_COMPILE
+#if defined(CORERT)
+EXTERN_C REDHAWK_API void* REDHAWK_CALLCONV RhpHandleAlloc(void* pObject, int type);
+EXTERN_C REDHAWK_API void REDHAWK_CALLCONV RhHandleFree(void*);
+EXTERN_C void* GetModuleSection(int id, int* length);
+EXTERN_C Object* RhNewObject(PTR_EEType);
+
+// API to obtain the thread static field
+COOP_PINVOKE_HELPER(Object*, RhGetThreadStaticField, (void* pModuleFieldTypePtr))
+{
+    int length = 0;
+    void* start = GetModuleSection(3 /* ThreadStaticRegionStart */, &length);
+    return ThreadStore::RawGetCurrentThread()->GetThreadStaticField((void**) pModuleFieldTypePtr - (void**) start);
+}
+
+Object* Thread::GetThreadStaticField(int offset)
+{
+    return *((Object**) m_pThreadStaticBase[offset]);
+}
+
+// Eagerly construct thread statics from the module's
+// thread static region containing the EEType.
+void Thread::ConstructThreadStatics()
+{
+    if (m_pThreadStaticBase != NULL)
+    {
+        return;
+    }
+
+    int byteLength = 0;
+    PTR_EEType* start = (PTR_EEType*) GetModuleSection(3 /* ThreadStaticRegionStart */, &byteLength);
+    
+    int length = byteLength / sizeof(void*);
+    void** pThreadStaticBase = new (nothrow) void*[length];
+    for (int i = 0; i < length; ++i)
+    {
+        Object* gcBlock = RhNewObject(start[i]);
+        pThreadStaticBase[i] = RhpHandleAlloc(gcBlock, 2 /* Normal */);
+    }
+
+    m_pThreadStaticBase = pThreadStaticBase;
+}
+
+// When the thread is destroyed, destroy its thread statics as well.
+void Thread::DestroyThreadStatics()
+{
+    if (m_pThreadStaticBase == NULL)
+    {
+        return;
+    }
+
+    int byteLength = 0;
+    (void) GetModuleSection(3 /* ThreadStaticRegionStart */, &byteLength);
```

Comment:
The section numbers will change in my check-in to now be based off 200 and up to mesh with the ReadyToRun spec just as a head's up.